singlem/pipe.py

```diff
@@ -71,6 +72,70 @@ def run(self, **kwargs):
+        #     def get_all_modes(tax):
+        #         count = Counter(tax)
+        #         max_count = max(count.values())
+        #         most_common = [(k, v) for k, v in count.items() if v == max_count]
+        #         return most_common
+
+        #     otu_table_collection = StreamingOtuTableCollection()
+        #     otu_table_collection.add_archive_otu_table_object(otu_table_object)
+
+        #     for _, sample_otus in otu_table_collection.each_sample_otus(generate_archive_otu_table=True):
+            
+        #         tax_per_read = {}
+        #         for otu in sample_otus:
+        #             if otu.taxonomy_assignment_method() == QUERY_BASED_ASSIGNMENT_METHOD:
+        #                 for read in otu.read_names():
+        #                     real_read = read.split('~')[0]
+        #                     if real_read not in tax_per_read:
+        #                         tax_per_read[real_read] = otu.equal_best_hit_taxonomies()
+        #                     else:
+        #                         tax_per_read[real_read] += otu.equal_best_hit_taxonomies()
+        #                     # tax_per_read[real_read] = list(itertools.chain.from_iterable(tax_per_read[real_read]))
+                
+        #         # getting the mode for each read
+        #         # might be good to add a check to make sure the mode
+        #         read_modes = {}
+        #         for real_read, tax in tax_per_read.items():
+        #             # print(real_read)
+        #             total = len(tax)
+        #             modes = get_all_modes(tax)
+        #             # if len(modes) > 1:
+        #                 # print("Multiple modes found: {}".format(modes))
+        #             # else:
+        #                 # print("Mode: {}\nAccounting for {} hits.".format(modes[0][0], modes[0][1]/total))
+        #             if len(modes) == 1:
+        #                 read_modes[real_read] = {'taxon': modes[0][0], 'fraction': modes[0][1]/total}
+
+        #         # sort read_modes so that the most common reads are last but just keep the taxon
+        #         read_modes = {k: v['taxon'] for k, v in sorted(read_modes.items(), key=lambda item: item[1]['fraction'])}
+
+        #         # reassign taxonomies of reads with modes to match the mode
+        #         for otu in sample_otus:
+        #             for read in otu.read_names():
+        #                 real_read = read.split('~')[0]
+        #                 if real_read in read_modes:
+        #                     otu.data[ArchiveOtuTable.TAXONOMY_FIELD_INDEX] = read_modes[real_read].replace(' ', '').replace(';', '; ')
+        #                     otu.data[ArchiveOtuTable.EQUAL_BEST_HIT_TAXONOMIES_INDEX] = [read_modes[real_read]]
+        #                     otu.data[ArchiveOtuTable.TAXONOMY_ASSIGNMENT_METHOD_INDEX] = QUERY_BASED_ASSIGNMENT_METHOD
+
+        # collapse_by_read(otu_table_object)
+
```

Comment:
I think this code is best removed from the already massive pipe file. Instead copy it into your notes or a branch or something, and if needed we can grab it from there. In my experience code like this is rarely useful in future.